en/lessons/basics/collections.md

```diff
@@ -3,209 +3,774 @@ version: 1.3.1
+operations on maps, and they are just as performant! Lists also have the set
+operations but they are much slower due to the different performance
+characteristics of linked lists and binary trees.
+
+## Adding Values to a Map
+
+The blessed function for adding to a map is called `insert`. If a duplicate key
+is added to a map, it will replace the former value:
+
+```haskell
+ghci> empty
+fromList []
+ghci> oneItem = insert "first" 1 empty
+ghci> oneItem
+fromList [("first",1)]
+ghci> insert "first" 2 oneItem
+fromList [("first",2)]
 ```
 
-**Note**: this syntax only works for updating a key that already exists in the map! If the key does not exist, a `KeyError` will be raised.
+A useful function to be aware of is `adjust`. This lets us update a value at a
+specified key, only if it exists, if not the old map is returned.
 
-To create a new key, instead use [`Map.put/3`](https://hexdocs.pm/elixir/Map.html#put/3)
+```haskell
+ghci> adjust (+2) "first" oneItem
+fromList [("first",3)]
+ghci> :t adjust
+adjust :: Ord k => (a -> a) -> k -> Map k a -> Map k a
+ghci> adjust (+2) "second" oneItem
+fromList [("first",1)]
+```
 
-```elixir
-iex> map = %{hello: "world"}
-%{hello: "world"}
-iex> %{map | foo: "baz"}
-** (KeyError) key :foo not found in: %{hello: "world"}
-    (stdlib) :maps.update(:foo, "baz", %{hello: "world"})
-    (stdlib) erl_eval.erl:259: anonymous fn/2 in :erl_eval.expr/5
-    (stdlib) lists.erl:1263: :lists.foldl/3
-iex> Map.put(map, :foo, "baz")
-%{foo: "baz", hello: "world"}
+The observant reader will notice that `adjust` doesn't actually update the map.
+The second invocation of adjust returns the original `oneItem` map, this makes
+sense when you consider that all data in haskell is immutable!
+
+## When to use Maps
+
+Maps are really great for in memory persistence of state that will need to be
+retrieved by a key of some arbitrary type. This is is because maps have great
+lookup asymptotics (`O(log n)`) due to the ordering on the key values. The
+example that immediately springs to mind is session storage on the server in a
+web application. The session state can be indexed by an id that is stored in
+the cookie, and the session state can be retrieved from an in memory `Map` on
+every request. This solution doesn't scale infinitely, but you would be
+surprised how well it works!
+
+## Hashmaps
+
+There are some cases where we do not have an ordering on our type, but still
+want to use it as a key to index a map. In this case we probably want to reach
+for a hashmap. A hashmap simply hashes the key, et voilà, we have an ordering!
+This does require that the key type is hashable though.
+
+The module that exports the `HashMap` data type and functionality is called
+`Data.HashMap.Strict`, it lives in the `unordered-containers` package.
+Unfortunately `unordered-containers` are not included in ghci (as of 9.0.1) so
+you will have to setup a project in order to pull in that dependency. Fret not
```

Comment:
You can remove this section, using external packages on-the-fly is covered by @soupi [here](https://github.com/haskell-infra/www.haskell.org/pull/88/files#diff-6e06515857e7603e2d734c3a97cde77cf7d00344596c99c8d257e98e5526b49aR236), and I intend to link to this page at the beginning of the Basics track.